storage/innobase/trx/trx0purge.cc

```diff
@@ -428,16 +443,25 @@ trx_purge_truncate_rseg_history(trx_rseg_t &rseg,
   const trx_id_t undo_trx_no=
     mach_read_from_8(b->page.frame + hdr_addr.boffset + TRX_UNDO_TRX_NO);
 
-  if (undo_trx_no >= limit.trx_no)
+  if (undo_trx_no >= trx_no)
   {
-    if (undo_trx_no == limit.trx_no)
-      err = trx_undo_truncate_start(&rseg, hdr_addr.page,
-                                    hdr_addr.boffset, limit.undo_no);
+    if (undo_trx_no == trx_no)
+      err= trx_undo_truncate_start(&rseg, hdr_addr.page,
+                                   hdr_addr.boffset, undo_no);
     goto func_exit;
   }
-
-  if (!all)
-    goto func_exit;
+  else
+  {
+    rseg_ref= rseg.ref_load();
+    if (rseg_ref >= rseg.REF || !purge_sys.sees(rseg.needs_purge))
```

Comment:
The MySQL 5.7 design may be affected by a race condition that was fixed in 0de3be8cfdfc26f5c236eaefe12d03c7b4af22c8. As far as I remember, the condition on `rseg_ref` is needed because `trx_assign_rseg_low()` may successfully invoke `rseg->acquire_if_available()` while we are holding the exclusive `rseg->latch`, and later expect some fields in `rseg` not to change by the time it invokes `trx_undo_assign()`. The `rseg` code has been refactored to reduce contention on the former `rseg->mutex`. It was replaced with some atomic memory access and `rseg->latch`.

You are right that a long running transaction could block this optimization. In that case, 126 or 127 other rollback segments would remain available for newly started transactions. In MySQL 5.7, only 96 rollback segments are available for persistent undo log records, or 95 if `innodb_undo_tablespaces>0`. I do not think that losing one of 127 or 128 rollback segments for some time is a big deal, and the performance tests that were run on this seem to agree. In the worst case (`innodb_undo_log_truncate=ON` and `innodb_undo_tablespaces=2`), we should have at least 62 rollback segments available for logging new transactions. The first rollback segment in the system tablespace would be unused, up to 64 rollback segments would be disabled by the undo tablespace truncation, and additionally one rollback segment in the not-being-truncated undo tablespace would be disabled by this new logic.

Reply:
Ok, thanks, I will have a look at the race condition that was fixed in https://github.com/MariaDB/server/commit/0de3be8cfdfc26f5c236eaefe12d03c7b4af22c8.